cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -188,65 +188,71 @@ impl YaccParser {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicateAvoidInsertDeclaration(spans.clone()),
                 span: *orig_span,
-            });
+            }]);
         }
         if let Some((orig_span, spans)) = self.duplicate_precedence_spans.iter().next() {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicatePrecedence(spans.clone()),
                 span: *orig_span,
-            });
+            }]);
         }
         if let Some((orig_span, spans)) = self.duplicate_implicit_token_spans.iter().next() {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicateImplicitTokensDeclaration(spans.clone()),
                 span: *orig_span,
-            });
+            }]);
         }
         if let Some((orig_span, spans)) = &self.duplicate_expect_declarations {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicateExpectDeclaration(spans.clone()),
                 span: *orig_span,
-            });
+            }]);
         }
         if let Some((orig_span, spans)) = &self.duplicate_expectrr_declarations {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicateExpectRRDeclaration(spans.clone()),
                 span: *orig_span,
-            });
+            }]);
         }
         if let Some((orig_span, spans)) = &self.duplicate_start_declarations {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicateStartDeclaration(spans.clone()),
                 span: *orig_span,
-            });
+            }]);
         }
         if let Some((orig_span, spans)) = self.duplicate_epp_declarations.iter().next() {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicateEPP(spans.clone()),
                 span: *orig_span,
-            });
+            }]);
         }
         if let Some(spans) = &self.duplicate_actiontype_declarations {
-            return Err(YaccGrammarError {
+            return Err(vec![YaccGrammarError {
                 kind: YaccGrammarErrorKind::DuplicateActiontypeDeclaration(spans.clone()),
                 span: *self
                     .global_actiontype
                     .as_ref()
                     .map(|(_, span)| span)
                     .unwrap(),
-            });
+            }]);
         }
-        i = self.parse_rules(i)?;
-        self.parse_programs(i)
+        i = self.parse_rules(i?).map_err(|e| vec![e]);
```

Comment:
I think this chunk has been moved to the wrong place? It looks like it should be at line 483?

Reply:
this is "the creation of DuplicateRule errors was moved from parse_rules into parse." part of the commit message,

This movement should be entirely equivalent to what was there, since we move it from right before returning `Ok`,
to right before we check the result returned by the function it is moved from.

There are 2 reasons for it:
1. It allows us to keep all of these checks in the same place.
2. Allows us to avoid switching the return type of `parse_rules`, and allocating a temporary vector for it in the next patch.

Reply:
Ah, I get it.